### PR TITLE
(Closes #457) Fix bug with backslash in strings

### LINKED ISSUE #457: Error on concatenation of two strings with a backslash
The following minimal program results in a `NoMatchError` for line 2:

```fortran
program simple
    character(len=10) :: foo = '\' // 'b'
end program simple
```

I believe this is correct Fortran--at least it compiles fine on Intel.

I'm still investigating myself. It's doesn't seem to be a problem with the regex pattern for the `concat-op`. I think it has something to do with the fact that backslashes in strings get automatically escaped to `\\` in Python.

Interestingly, if I take out the line from the program block, i.e. just:
```fortran
character(len=10) :: foo = '\' // 'b'
```
it works fine.

### PR DESCRIPTION
**Pull request by arporter**
Weirdly, fparser was interpreting backslash `"\"` as an escape character. This is not a standard Fortran feature. i.e. if you want to escape a quotation mark in a Fortran string you simply repeat it: `"''"` means `"'"`.

### CODE REVIEW

File: src/fparser/common/splitline.py [hunk 1]
```diff
@@ -249,94 +250,58 @@ def splitquote(line, stopchar=None, lower=False, quotechars="\"'"):
...
-        if part:
-            item = String("".join(part))
-            items.append(item)
-    return items, stopchar
+                segment = String(line[start:i])
+                segments.append(segment)
+                quote_char = None
+        else:
+            start = i
+            while i < n and (quote_char is not None or line[i] not in ("'", '"')):
```
Comment (hiker) [inline src/fparser/common/splitline.py:297]: I am not really convinced that this is correct (and if I am wrong, it needs some comments):
if the caller sets `stop_char` (which is what `quote_char` is initialised with), then I believe this while loop becomes basically:
```
while i < n and True:
```
i.e. it will just return the rest of the string, ignoring any potential `stop_char` in the string?? I also didn't see a test that provides the `stop_char` to the call.

File: src/fparser/common/splitline.py [hunk 2]
```diff
@@ -249,94 +250,58 @@ def splitquote(line, stopchar=None, lower=False, quotechars="\"'"):
...
-                    # This marks the end of the current part.
-                    break
-                if char == "\\":
-                    nofslashes += 1
+    segments = []
+    i = 0
+    n = len(line)
+    quote_char = stopchar
+
+    while i < n:
```
Comment (hiker) [inline src/fparser/common/splitline.py:270]: A comment or two would be useful:
`A quote character was found, collect all characters till closing quote, but ignore two consecutive quote characters`
OK, that's awful ;) But something like this.

File: src/fparser/common/splitline.py [hunk 3]
```diff
@@ -249,94 +250,58 @@ def splitquote(line, stopchar=None, lower=False, quotechars="\"'"):
...
+        if quote_char is None and line[i] in ("'", '"'):
+            quote_char = line[i]
+            start = i
+            i += 1
+            while i < n:
+                if line[i] == quote_char:
+                    if i + 1 < n and line[i + 1] == quote_char:
```
Comment (hiker) [inline src/fparser/common/splitline.py:277]: A comment "Ignore two consecutive quote characters"

File: src/fparser/common/splitline.py [hunk 4]
```diff
@@ -249,94 +250,58 @@ def splitquote(line, stopchar=None, lower=False, quotechars="\"'"):
     current closing quotation character to be specified.
```
Comment (hiker) [inline src/fparser/common/splitline.py:250]: Add a comment that explains that a double-quote-character is an 'escaped quote'.

File: src/fparser/common/tests/test_splitline.py [hunk 5]
```diff
@@ -165,22 +165,76 @@ def test_splitparen():
...
+        # Quoted string at end
+        ('label = "xyz"', ["label = ", '"xyz"'], None),
+        # Embedded commas
+        ('DATA STR /"A,B,C"/', ["DATA STR /", '"A,B,C"', "/"], None),
+        # Fortran character kind (should treat as unquoted)
+        ("character(len=5, kind=1) :: foo", ["character(len=5, kind=1) :: foo"], None),
+        # Unterminated double-quoted string at end
+        ('PRINT *, "unterminated', ["PRINT *, ", '"unterminated'], '"'),
+        # Unterminated single-quoted string at end
+        ("PRINT *, 'unterminated", ["PRINT *, ", "'unterminated"], "'"),
```
Comment (hiker) [inline src/fparser/common/tests/test_splitline.py:206]: Can you also add a test for an empty string, e.g. `print *,"test", "", '', "the end"`

I originally thought your new loop would not work, but it does ;)

File: src/fparser/common/tests/test_splitline.py [hunk 6]
```diff
@@ -165,22 +165,76 @@ def test_splitparen():
...
+def test_split_fortran_strings(input_line, expected_parts, expected_unterm):
+    parts, unterminated = splitquote(input_line)
+    assert parts == expected_parts, (
+        f"For input: {input_line!r} got parts: {parts!r} but expected: "
+        f"{expected_parts!r}"
+    )
+    assert unterminated == expected_unterm, (
+        f"For input: {input_line!r} got unterminated: {unterminated!r} but "
+        f"expected: {expected_unterm!r}"
+    )
```
Comment (hiker) [inline src/fparser/common/tests/test_splitline.py:237]: Please also add a test when a quote character is passed into `splitquote`.

**Review by hiker — CHANGES_REQUESTED** [review]
Thanks a lot for this. The loop looks a lot simpler that what it was before, that's great. I do have the feeling that there is a bug though, please check the comments (and if I am wrong, it needs some tests :) ).

File: src/fparser/common/splitline.py (hunk 1 above)
Comment (arporter) [reply src/fparser/common/splitline.py:297]: Darn, you are right. So much for AI.

File: src/fparser/common/splitline.py [hunk 7]
```diff
@@ -239,7 +238,48 @@ def string_replace_map(line, lower=False):
     return "".join(items), string_map
 
 
-def splitquote(line, stopchar=None, lower=False, quotechars="\"'"):
+def _next_quote(line: str, quote_char: Optional[str] = None, start: int = 0) -> int:
+    """
```
Comment (hiker) [inline src/fparser/common/splitline.py:242]: Nice, that's easier to understand now.

File: src/fparser/common/splitline.py (hunk 4 above)
Comment (hiker) [reply src/fparser/common/splitline.py:289]: Thanks!

File: src/fparser/common/tests/test_splitline.py (hunk 5 above)
Comment (hiker) [reply src/fparser/common/tests/test_splitline.py:217]: Thanks!

File: src/fparser/common/splitline.py (hunk 1 above)
Comment (hiker) [reply src/fparser/common/splitline.py]: Human: 1, AI: 0

**Review by hiker — APPROVED** [review]
Ready to be merged, thanks!
